messages/jsonschema/ParameterType.json

```diff
@@ -29,6 +28,12 @@
     },
     "id": {
       "type": "string"
+    },
+    "children": {
+      "items": {
+        "$ref": "#"
```

Comment:
This seems to have broken the generated TypeScript code (see CI failure). I assume ˋ#ˋ means self-reference? Where in the json schema spec is that documented?

Reply:
I haven't found any mentions of `#` meaning self-reference. [This](https://stackoverflow.com/questions/35250621/recursive-self-referencing-json-schema) SO suggests the reference needs to go via `definitions`.

And [this](https://stackoverflow.com/questions/20752716/json-schema-recursive-schema-definition/20753545#20753545) post suggests that `#` refers to the root of the document (not "self")

Reply:
Well, what _is_ the correct type for an item in the `children` array? We assumed it was instances of `ParameterType` so that `#` is meant to refer to the root of the document. Maybe `./ParameterType.json` would be better?

Reply:
There's some info about JSON pointers [here](https://json-schema.org/understanding-json-schema/structuring.html#id1) and a [spec](https://datatracker.ietf.org/doc/html/rfc6901) which talks about `#` referring to "the whole document".

Reply:
> Well, what is the correct type for an item in the children array?

Sorry, my bad - I saw `children` and incorrectly assumed this: https://github.com/cucumber/common/blob/1b64380f36cf6bb8534b59af7b064e84299c7b22/messages/messages.proto#L559-L563

Which in JSON Schema is this:

https://github.com/cucumber/common/blob/1b64380f36cf6bb8534b59af7b064e84299c7b22/messages/jsonschema/TestCase.json#L5-L28

The `ParameterType` message doesn't have a `children` property at all:

https://github.com/cucumber/common/blob/1b64380f36cf6bb8534b59af7b064e84299c7b22/messages/messages.proto#L681-L688

A bit of background: I used a `.proto` to `JSON Schema` conversion tool to generate the JSON schema files. The tool didn't convert everything correctly, so I've done many manual edits to the generated schemas. In the process I've apparently made several mistakes (copy paste errors) in the `required` sections. There are probably more.

These mistakes have gone unnoticed until now, because the custom Ruby code generator doesn't generate code based on the `required` array, but rather the `properties` dictionary of the schema:

* https://github.com/cucumber/common/blob/1b64380f36cf6bb8534b59af7b064e84299c7b22/messages/jsonschema/scripts/codegen.rb#L115
* https://github.com/cucumber/common/blob/1b64380f36cf6bb8534b59af7b064e84299c7b22/messages/jsonschema/scripts/codegen.rb#L164

This means mistakes/inconsistencies in `required` are not detected. We should probably change the code generators to iterate over `required` instead. That way we should be able to find the remaining mistakes.

Sorry again for leading you down the wrong path!

Reply:
While you're working on the schemas - could you also please replace all the `"$schema": "http://json-schema.org/draft-04/schema#"` with `"$schema": "https://json-schema.org/draft/2020-12/schema`? (The conversion tool I used put those old versions in).

FWIW, there isn't currently a non-draft version of the JSON schema spec - see [available versions](https://json-schema.org/specification-links.html)).